### ISSUE TITLE
Add support for GEOMETRY type

### ISSUE THREAD

**Issue by gmullerb** [r2dbc/r2dbc-h2#80]
Currently there is no support for spatial geometry type:

- http://www.h2database.com/html/datatypes.html#geometry_type
- https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/value/ValueGeometry.java

In the meantime, anyone having same requirement can:
```sql
CREATE TABLE theTable (
  location GEOMETRY(POINT) NOT NULL,
);
```

```java
final WKTReader wktReader = new WKTReader();
r2dbc.inTransaction(handle -> handle
  .select("SELECT CAST(location AS VARCHAR) AS loc FROM theTable")
  .mapRow(row -> row.get("loc", String.class))
  .flatMap(loc -> Mono.create(sink -> {
    try {
      sink.success(wktReader.read(loc));
    }
    catch (Exception ex) {
      sink.error(ex);
    }
  })));
```

**Comment by katzyn** [comment]
H2 does not have a fixed default mapping for this data type in recent versions. H2 returns `org.locationtech.jts.geom.Geometry` only when JTS library is in classpath for H2GIS and for compatibility with own older versions. H2 returns a `String` with EWKT representation of geometry when this third-party library wasn't found. R2DBC-H2 should handle it somehow and R2DBC-H2 should not have strong dependencies on JTS.

Of course, you can use `org.locationtech.jts.io.WKTReader` in your own code, but R2DBC-H2 should not use it. `WKTReader` can't always read EWKT produced by H2, it can read only a subset of it. Also the double conversion to EWKT and to `Geometry` is slow. H2's own method should be used instead.

**Comment by gmullerb** [comment]
Exactly, using `org.locationtech.jts.io.WKTReader` is really an agony, too much code for that, having a Codec for `Geometry` could be a good solution, independent of what use is given, besides, as you said, `h2` already depends on `org.locationtech.jts:jts-core` [h2/pom.xml](https://github.com/katzyn/h2database/blob/master/h2/pom.xml) and `r2dbc-h2` depends on `com.h2database:h2` [r2dbc-h2/pom.xml](https://github.com/r2dbc/r2dbc-h2/blob/master/pom.xml). Something like this [commit 2a14639](https://github.com/gmullerb/r2dbc-h2/commit/2a14639c8bb3ca0033825b79e6748eee4aded040)

**Comment by katzyn** [comment]
Please, don't use `pom.xml` as a reference document (and don't use it to build H2 from the sources). `GEOMETRY` data type in H2 does not really depend on JTS since 1.4.198. Actually H2 never had any strong dependencies.

JTS library is **optionally** used to read passed `org.locationtech.jts.geom.Geometry` objects and to return such objects. Hypothetically H2 may support other libraries in the future, but AFAIK nobody asked for such support yet.

However, anyone can use (E)WKB (as `byte[]`) or (E)WKT (as `String`) representations as is or together with some other geometry library. R2DBC-H2 should also be able to work with `GEOMETRY` data type without JTS. But R2DBC-H2 may (and probably should) be able to support `org.locationtech.jts.geom.Geometry` when JTS library is in classpath.

**Comment by katzyn** [comment]
With JDBC there is a way to specify the required data type.
```Java
ResultSet rs = …;
Geometry geometry = rs.getObject(1, Geometry.class);
byte[] ewkb = rs.getObject(1, byte[].class); // or getBytes(1);
String ewkt = rs.getObject(1, String.class); // or getString(1);
```

Simple `getObject(1)` returns either `Geometry` or `String` depending on JTS presence in the classpath of H2.

R2DBC-H2 should probably also support all 3 supported by H2 formats to read `GEOMETRY` objects from H2 and to pass them to H2.

**Comment by gmullerb** [comment]
Using JDBC will circumvent the whole purpose of using Reactive programming, the purpose of using r2dbc-h2 at the first place. In [commit 2a14639](https://github.com/gmullerb/r2dbc-h2/commit/2a14639c8bb3ca0033825b79e6748eee4aded040), I can make the `org.locationtech.jts:jts-core` dependency `provided`, so it can be "optional". (I tested  and it is working good)

**Comment by katzyn** [comment]
In your branch it isn't optional, because `GeometryCodec` is always instantiated and its instantiation uses a direct reference to `org.locationtech.jts.geom.Geometry` class. You need to check JTS presence before instantiation of `GeometryCodec`.

Also your branch does not provide any normal way to read geometries in absence of JTS.

I'm not a R2DBC reviewer, so you need to discuss your work with R2DBC community.

**Comment by gmullerb** [comment]
Yes, right now it not optional, I will work on that as soon as I can and make a pull request proposal, Thank You So Much!

**Comment by gregturn** [comment]
I’m willing to have a codec that optionally registers itself based on classpath scanning.

**Comment by gmullerb** [comment]
Superb!

**Comment by gmullerb** [comment]
Done: https://github.com/r2dbc/r2dbc-h2/pull/81
Thanks!
